### PR TITLE
raftstore: only transfer leader to active peers

### PR DESCRIPTION
**Pull request by hicqu**
Signed-off-by: qupeng <qupeng@pingcap.com>

## What have you changed? (mandatory)
Only transfer leadership to active peers.

In this PR we add a check in leader transferring. If the target peer is not active, the transferring will be skipped.

## What are the type of the changes? (mandatory)
Improvement.

## How has this PR been tested? (mandatory)
CI.

## Does this PR affect documentation (docs) or release note? (mandatory)

### CODE REVIEW

**Comment by dbjoa** [comment]
@hicqu 
Is the active peer chosen by the PreVote?

**Comment by siddontang** [comment]
@hicqu 

can we add some tests?

**Comment by hicqu** [comment]
@dbjoa in Raft leader sends heartbeats to followers and receives heartbeat responses from them. If leader has not received responses from one follwer in an election time, the follower will be treated as inactive by its leader.

**Comment by dbjoa** [comment]
@hicqu 
If we already know the likelihood of certain members of a Raft group becoming a Leader via [Raft PreVote](https://pingcap.com/blog/tidb-2.1-ga-Battle-tested-to-handle-an-unpredictable-world/), we can try to transfer the leadership first to one of the most likely candidates without sending or waiting these heartbeats. Isn't it? Would you explain why you don't exploit the likelihood information computed in advance?

**Comment by hicqu** [comment]
@dbjoa `PreVote` is not designed for leader transferring. If leader transfers the leadership to the peer who sends `PreVote`, there could be such issues:
1. if the peer is just isolated and then re-joins the raft group, it will introduce useless leader transferring;
2. after the node is restarted, all raft peers on the store could become leader after they send `PreVote` to their leader; I think it's not what we want.

The PR introduces this behavior to TiKVs:
1. after the node is restarted, it tells PD "Hi, I'm back";
2. after that, PD trys to transfer some peer leaderships to the store;
3. for every peer, if the leader knows the target is not active, the proposal will be dropped;
  without the PR, the leader transferring will be proposed, and needs more time to finish.
  during the time, new proposals will be dropped, so clients' write will fail.

File: components/test_raftstore/src/transport_simulate.rs [hunk 1]
```diff
@@ -753,3 +753,23 @@ impl Filter for DropMessageFilter {
         Ok(())
     }
 }
+
+#[derive(Clone)]
+pub struct MessageCorruptFilter<F: Fn(&mut RaftMessage) + Clone> {
```
Comment (lonng) [inline components/test_raftstore/src/transport_simulate.rs]: Prefer to `MessageCorruptFilter<F: Fn(&mut RaftMessage) + Clone>(F)` and remove the `new` method.

File: components/test_raftstore/src/transport_simulate.rs [hunk 2]
```diff
@@ -753,3 +753,23 @@ impl Filter for DropMessageFilter {
...
+    f: F,
+}
+
+impl<F: Fn(&mut RaftMessage) + Clone> MessageCorruptFilter<F> {
+    pub fn new(f: F) -> Self {
+        MessageCorruptFilter { f }
+    }
+}
+
+impl<F: Fn(&mut RaftMessage) + Clone + Send + Sync> Filter for MessageCorruptFilter<F> {
```
Comment (lonng) [inline components/test_raftstore/src/transport_simulate.rs]: The `F` contains so many contracts, why not move them to where clause.

**Comment by hicqu** [comment]
PTAL @BusyJay @overvenus

File: tests/integrations/raftstore/test_multi.rs [hunk 3]
```diff
@@ -351,11 +352,14 @@ fn test_leader_change_with_uncommitted_log<T: Simulator>(cluster: &mut Cluster<T
     ));
     // Make peer 2 have no way to know the uncommitted entries can be applied
     // when it's still follower.
-    cluster.add_send_filter(CloneFilterFactory(
-        RegionPacketFilter::new(1, 2)
-            .msg_type(MessageType::MsgHeartbeat)
-            .direction(Direction::Recv),
-    ));
+    cluster.add_send_filter(CloneFilterFactory(MessageCorruptFilter(
```
Comment (overvenus) [inline tests/integrations/raftstore/test_multi.rs]: Why touch this test?

Comment (hicqu) [reply tests/integrations/raftstore/test_multi.rs]: The test case blocks heartbeats from leader to a follower, and then transfers leadership to that peer. This behavior is broken by the PR.

File: src/raftstore/store/peer.rs [hunk 4]
```diff
@@ -1833,6 +1833,27 @@ impl Peer {
...
+                "peer_id" => self.peer.get_id(),
+                "peer" => ?peer,
+            );
+            return false;
+        }
+
+        let pr = progress.get(peer_id).unwrap();
+        if !pr.recent_active || pr.state != ProgressState::Replicate {
+            // After the target store is reported as unreachable,
+            // the state will become ProgressState::Probe.
```
Comment (BusyJay) [inline src/raftstore/store/peer.rs]: Is it possible that the state never becomes Replicate?

**Comment by hicqu** [comment]
@BusyJay after a TiKV store restarts, it reports to PD so that PD will schedule some leaders to that store. However maybe not all peers on that store are ready, so if some region sends `MsgTimeoutNow` to the peer, the peer can't become leader immediately.

**Pull request closed without merging**